### PR TITLE
liteclient_test: extend test suite

### PR DESCRIPTION
**Pull request by ldts**
The following tests have been added:

 OstreeUpdateWhenNoInstalledVersions
   perform an update when the initial version is not available

 OstreeUpdateInstalledVersionsCorrupted1
   invalidate the installed_verion json by corrupting the sysroot hash

 OstreeUpdateInstalledVersionsCorrupted2
   write an invalid json to the filesystem

Also refactored the code a little bit to improve its readability.

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

### CODE REVIEW

File: tests/liteclient_test.cc [hunk 1]
```diff
@@ -220,19 +250,17 @@ class LiteClientTest : public ::testing::Test {
...
-    target_json["custom"]["targetFormat"] = "OSTREE";
-    target_json["length"] = 0;
-    initial_target_ = Uptane::Target{hw_id + "-" + os + "-1", target_json};
+                    initial_target_{Uptane::Target::Unknown()},
+                    sysroot_hash_{sysrepo_.repo().commit(sysrootfs_.path, sysrootfs_.branch)} {
+    sysrepo_.deploy(sysroot_hash_);
   }
 
-  std::shared_ptr<LiteClient> create_liteclient(bool initial_version = true) {
+  enum initial_version { off = 0, on = 1, corrupted1 = 2, corrupted2 = 3 };
```
Comment (mike-sul) [inline tests/liteclient_test.cc:258]: Please, follow the google C++ code guideline (e.g. off -> kOff).

File: tests/liteclient_test.cc [hunk 2]
```diff
@@ -220,19 +250,17 @@ class LiteClientTest : public ::testing::Test {
...
+
+  /**
+   * method create_liteclient
+   */
+  std::shared_ptr<LiteClient> create_liteclient(enum initial_version initial_version = initial_version::on) {
```
Comment (mike-sul) [inline tests/liteclient_test.cc:263]: 'enum' keyword is redundant here.

**Comment by mike-sul** [comment]
Nice work! Just minor comments.

**Comment by ldts** [comment]
@mike-sul updated PR

File: tests/liteclient_test.cc [hunk 3]
```diff
@@ -220,19 +250,17 @@ class LiteClientTest : public ::testing::Test {
...
-    target_json["custom"]["targetFormat"] = "OSTREE";
-    target_json["length"] = 0;
-    initial_target_ = Uptane::Target{hw_id + "-" + os + "-1", target_json};
+                    initial_target_{Uptane::Target::Unknown()},
+                    sysroot_hash_{sysrepo_.repo().commit(sysrootfs_.path, sysrootfs_.branch)} {
+    sysrepo_.deploy(sysroot_hash_);
   }
 
-  std::shared_ptr<LiteClient> create_liteclient(bool initial_version = true) {
+  enum initial_version { kOff = 0, kOn = 1, kCorrupted1 = 2, kCorrupted2 = 3 };
```
Comment (mike-sul) [inline tests/liteclient_test.cc:258]: https://google.github.io/styleguide/cppguide.html#Type_Names (--> enum class InitialVersion).

Comment (ldts) [reply tests/liteclient_test.cc]: going to cleanup the whole file...is full of inconsistencies

File: tests/liteclient_test.cc [hunk 4]
```diff
@@ -170,38 +172,41 @@ class TufRepoMock {
...
-                    url_{"http://localhost:" + port_},
-                    req_headers_file_{tuf_repo_mock_.path() + "/headers.json"},
-                    process_{RunCmd, "--port", port_, "--ostree", ostree_mock_.path(), "--tuf-repo", tuf_repo_mock_.path(), "--headers-file", req_headers_file_} {
+  DeviceGatewayMock(const OSTreeRepoMock& ostree, const TufRepoMock& tuf)
+      : ostree_{ostree},
+        tuf_{tuf},
+        port_{TestUtils::getFreePort()},
+        url_{"http://localhost:" + port_},
+        req_headers_file_{tuf_.getPath() + "/headers.json"},
+        process_{RunCmd,           "--port",         port_, "--ostree", ostree_.getPath(), "--tuf-repo", tuf_.getPath(),
```
Comment (mike-sul) [inline tests/liteclient_test.cc:190]: looks like additional empty spaces?

File: tests/liteclient_test.cc [hunk 5]
```diff
@@ -210,144 +215,235 @@ class DeviceGatewayMock {
...
-  std::shared_ptr<LiteClient> create_liteclient(bool initial_version = true) {
+  enum class InitialVersion { kOff, kOn, kCorrupted1, kCorrupted2 };
+
+  /**
+   * method createLiteClient
+   */
+  std::shared_ptr<LiteClient> createLiteClient(InitialVersion initial_version = InitialVersion::kOn) {
     Config conf;
-    conf.uptane.repo_server = deviceGateway().tufRepoUri();
+    conf.uptane.repo_server = getDeviceGateway().getTufRepoUri();
```
Comment (mike-sul) [inline tests/liteclient_test.cc:245]: there is access to `device_gateway_` from this method.

File: tests/liteclient_test.cc [hunk 6]
```diff
@@ -210,144 +215,235 @@ class DeviceGatewayMock {
...
     conf.provision.primary_ecu_hardware_id = hw_id;
     conf.storage.path = test_dir_.Path();
 
     conf.pacman.type = ComposeAppManager::Name;
-    conf.pacman.sysroot = sysrepo_.path();
+    conf.pacman.sysroot = sys_repo_.getPath();
```
Comment (mike-sul) [inline tests/liteclient_test.cc:250]: IMO, `sys_repo_.path()` looks better in terms of readability, and getPath() actually doesn't do any getting it's just an accessor to the object field. Anyway, it's not really important.

File: tests/liteclient_test.cc [hunk 7]
```diff
@@ -210,144 +215,235 @@ class DeviceGatewayMock {
...
     conf.pacman.os = os;
     conf.pacman.extra["booted"] = "0";
     conf.pacman.extra["compose_apps_root"] = (test_dir_.Path() / "compose-apps").string();
-    conf.pacman.ostree_server = deviceGateway().ostreeUri();
+    conf.pacman.ostree_server = getDeviceGateway().getOsTreeUri();
```
Comment (mike-sul) [inline tests/liteclient_test.cc:254]: direct access to `device_gateway_`

**Review by mike-sul — APPROVED** [review]

File: tests/liteclient_test.cc (hunk 4 above)
Comment (ldts) [reply tests/liteclient_test.cc:190]: I didnt do those...must have been the clang format. should I remove them?

File: tests/liteclient_test.cc (hunk 6 above)
Comment (ldts) [reply tests/liteclient_test.cc:250]: I understand that but then we should just make the member public...it we are hacking we should at least make it explicit IMO

File: tests/liteclient_test.cc (hunk 7 above)
Comment (ldts) [reply tests/liteclient_test.cc:254]: sure, seems all this refactoring is bring up to the surface all the current hacks

**Comment by ldts** [comment]
@mike-sul if you want I can continue refactoring....I mean, it is an iterative work...

**Comment by mike-sul** [comment]
> @mike-sul if you want I can continue refactoring....I mean, it is an iterative work...

no, let's merge it.

**Comment by ldts** [comment]
let me address these two last comments; should be quick

**Pull request merged**